### PR TITLE
api: Support assigning qualified qmlelem to a property

### PR DESCRIPTION
**Pull request by stephenmdangelo**
e.g.

import QtQuick 2.5 as QQ

Repeater {
  delegate: QQ.Rectangle {}
}

### CODE REVIEW

**Comment by ChALkeR** [comment]
Sorry for not looking at this earlier.

This shouldn't change the existing testcase (removing the test that we already have, i.e. `property Item item: Item {}`). Instead, it should introduce a new one.

In this specific case either _adding_ a new `import` line and adding some new properties to `Properties.qml` or adding an entirely new file would work.

**Comment by ChALkeR** [comment]
Also I think that this comment by @akreuzkamp: https://github.com/qmlweb/qmlweb-parser/pull/17#issuecomment-231526384 is also applicable to this PR.
Shouldn't the names be converted to proper AST? For example, see [/ImportsNamed.qml](https://github.com/qmlweb/qmlweb-parser/blob/master/tests/qml/ImportsNamed.qml) and [ImportsNamed.qml.json](https://github.com/qmlweb/qmlweb-parser/blob/master/tests/qml/ImportsNamed.qml.json) files.

File: tests/qml/Properties.qml.json [hunk 1]
```diff
@@ -5,13 +5,13 @@
...
       2.5,
-      "",
+      "QQ",
       true
     ]
   ],
   [
     "qmlelem",
-    "Rectangle",
+    "QQ.Rectangle",
```
Comment (ChALkeR) [inline tests/qml/Properties.qml.json]: This is not the proper format for `qmlelem`. See [ImportsNamed.qml.json#L27-L32](https://github.com/qmlweb/qmlweb-parser/blob/master/tests/qml/ImportsNamed.qml.json#L27-L32) for how qualified `qmlelem` should be encoded. You could also test with longer paths. e.g. `Abc.Def.Rectangle`.

File: tests/qml/Properties.qml.json [hunk 2]
```diff
@@ -131,12 +131,12 @@
           "stat",
           [
             "qmlelem",
-            "Item",
+            "QQ.Item",
```
Comment (ChALkeR) [inline tests/qml/Properties.qml.json]: This is not the proper format for `qmlelem`. See [ImportsNamed.qml.json#L27-L32](https://github.com/qmlweb/qmlweb-parser/blob/master/tests/qml/ImportsNamed.qml.json#L27-L32) for how qualified `qmlelem` should be encoded. You could also test with longer paths. e.g. `Abc.Def.Rectangle`.

File: tests/qml/Properties.qml [hunk 3]
```diff
@@ -12,5 +12,5 @@ Rectangle {
   property int number2: 10;
   property var foo: {}
   property var bar: []
-  property Item item: Item {}
+  property Item item: QQ.Item {}
```
Comment (ChALkeR) [inline tests/qml/Properties.qml]: This should be added as a new line, and should not replace the current `property Item item: Item {}` test.

File: tests/qml/Properties.qml [hunk 4]
```diff
@@ -1,6 +1,6 @@
-import QtQuick 2.5
+import QtQuick 2.5 as QQ
 
-Rectangle {
+QQ.Rectangle {
```
Comment (ChALkeR) [inline tests/qml/Properties.qml]: Unneeded change.

File: tests/qml/Properties.qml [hunk 5]
```diff
@@ -1,6 +1,6 @@
-import QtQuick 2.5
+import QtQuick 2.5 as QQ
```
Comment (ChALkeR) [inline tests/qml/Properties.qml]: The original line should be present to keep the existing testcase (see comment below).
Perhaps use some other import, not `QtQuick`, and add it to the second line?

File: tests/qml/Properties.qml.json (hunk 1 above)
Comment (ChALkeR) [inline tests/qml/Properties.qml.json]: This has changed and will be `["dot",  "QQ", "Rectangle"],` after a rebase.
Note that this line shouldn't even be present here, as `Rectangle` → `QQ.Rectangle` is an unrelated change on this line.

**Review by ChALkeR — CHANGES_REQUESTED** [review]
`qmlelem` format changed in 6e5bf1ca2eacb2758c45d5789499d8d0b9cfcb3c, and those strings are now parsed deeper.

This PR should adapt to those changes after a rebase.
Also, it shouldn't remove the existing testcase for `property Item item: Item {}`.

**Comment by ChALkeR** [comment]
@stephenmdangelo, sorry for not reviewing at this earlier. In general, I'm in favor of this PR (and thanks for doing this!), but it requires some changes before landing. See the review above.

**Comment by stephenmdangelo** [comment]
@ChALkeR So I've cleaned up the Properties.json test (as well, I added a change to `tape.js` to fix the tests failing if, say, I was editing in vi).

I think I'm generating the dot expression as you're suggesting.

Though I find it odd that the `expr` as returned by `maybe_assign` is of the form:

```
[
  "dot",
  "DOM",
  [
    "name",
    "DOM"
  ],
  "Div"
]
```
But a non-assigned `qmlelem` (as in `ImportsNamed.qml`) omits the `name` array:

```
[
   "dot",
   "DOM",
   "Div"
]
```

So I went with the latter form. Is this correct? If so, why don't we use the `name` format, or why does that format get returned by `maybe_assign`?

File: tests/tape.js [hunk 6]
```diff
@@ -14,6 +14,9 @@ function buildTree(dir, data) {
   const subdirs = [];
   for (const file of fs.readdirSync(dir)) {
     const filePath = path.join(dir, file);
+    if (file[0] === ".") {
+        continue;
```
Comment (ChALkeR) [inline tests/tape.js:17]: Nit: should be above `const filePath`.

**Review by ChALkeR — APPROVED** [review]
LGTM with a nit.

**Comment by ChALkeR** [comment]
I will take a look at `maybe_assign` a bit later, though. =)

**Pull request merged**